File: apps/web/lib/heygen.ts
const HEYGEN_API_BASE = "https://api.heygen.com";

function getApiKey(): string {
  return process.env.HEYGEN_API_KEY || "";
}

function headers(): Record<string, string> {
  return {
    "X-Api-Key": getApiKey(),
    "Content-Type": "application/json",
  };
}

export interface HeyGenAvatar {
  avatar_id: string;
  avatar_name: string;
  preview_image_url: string;
  gender?: string;
  isCustom?: boolean;
}

// In-memory cache: avatar list + timestamp
let cachedAvatars: HeyGenAvatar[] | null = null;
let cacheTimestamp = 0;
const CACHE_TTL = 10 * 60 * 1000; // 10 minutes

/** List available HeyGen avatars (cached). */
export async function listAvatars(): Promise<HeyGenAvatar[]> {
  if (!getApiKey()) return [];

  // Return cached if fresh
  if (cachedAvatars && Date.now() - cacheTimestamp < CACHE_TTL) {
    return cachedAvatars;
  }

  const res = await fetch(`${HEYGEN_API_BASE}/v2/avatars`, {
    headers: headers(),
  });

  if (!res.ok) return cachedAvatars || [];

  const data = await res.json();
  // v2 returns { data: { avatars: [...] } }
  const allAvatars = data?.data?.avatars || [];
  cachedAvatars = allAvatars
    .filter((a: Record<string, unknown>) => a.preview_image_url)
    .map((a: Record<string, unknown>) => ({
      avatar_id: a.avatar_id as string,
      avatar_name: (a.avatar_name as string) || "Unnamed",
      preview_image_url: (a.preview_image_url as string) || "",
      gender: (a.gender as string) || undefined,
      // Custom avatars have hex-hash IDs; public ones use human-readable names
      isCustom: /^[0-9a-f]{32}$/i.test(a.avatar_id as string),
    }));
  cacheTimestamp = Date.now();

  return cachedAvatars!;
}

/** Get the preview photo URL for a given avatar ID. */
export async function getAvatarPhoto(avatarId: string): Promise<string | null> {
  const avatars = await listAvatars();
  const match = avatars.find((a) => a.avatar_id === avatarId);
  return match?.preview_image_url || null;
}

/** Create a streaming session token for the interactive avatar SDK. */
export async function createStreamingToken(): Promise<string | null> {
  if (!getApiKey()) return null;

  const res = await fetch(`${HEYGEN_API_BASE}/v1/streaming.create_token`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({}),
  });

  if (!res.ok) return null;

  const data = await res.json();
  return data?.data?.token || null;
}

File: apps/api/app/routers/heygen.py
"""HeyGen routes: /api/heygen/avatars, /api/heygen/token."""

from __future__ import annotations

import re
import time

import httpx
from fastapi import APIRouter, HTTPException, Depends

from ..deps import get_clerk_user_id
from ..config import get_settings

router = APIRouter(prefix="/api/heygen", tags=["heygen"])

HEYGEN_API_BASE = "https://api.heygen.com"

# In-memory cache
_cached_avatars: list[dict] | None = None
_cache_timestamp: float = 0
CACHE_TTL = 10 * 60  # 10 minutes


@router.get("/avatars")
async def list_avatars(user_id: str = Depends(get_clerk_user_id)):
    global _cached_avatars, _cache_timestamp

    settings = get_settings()
    if not settings.heygen_api_key:
        return {"avatars": []}

    if _cached_avatars is not None and time.time() - _cache_timestamp < CACHE_TTL:
        return {"avatars": _cached_avatars}

    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{HEYGEN_API_BASE}/v2/avatars",
            headers={"X-Api-Key": settings.heygen_api_key, "Content-Type": "application/json"},
            timeout=30,
        )

    if not resp.is_success:
        return {"avatars": _cached_avatars or []}

    data = resp.json()
    all_avatars = data.get("data", {}).get("avatars", [])

    _cached_avatars = [
        {
            "avatar_id": a["avatar_id"],
            "avatar_name": a.get("avatar_name", "Unnamed"),
            "preview_image_url": a.get("preview_image_url", ""),
            "gender": a.get("gender"),
            "isCustom": bool(re.match(r"^[0-9a-f]{32}$", a["avatar_id"], re.IGNORECASE)),
        }
        for a in all_avatars
        if a.get("preview_image_url")
    ]
    _cache_timestamp = time.time()

    return {"avatars": _cached_avatars}


@router.post("/token")
async def create_token(user_id: str = Depends(get_clerk_user_id)):
    settings = get_settings()
    if not settings.heygen_api_key:
        raise HTTPException(status_code=500, detail="failed_to_create_token")

    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{HEYGEN_API_BASE}/v1/streaming.create_token",
            headers={"X-Api-Key": settings.heygen_api_key, "Content-Type": "application/json"},
            json={},
            timeout=30,
        )

    if not resp.is_success:
        raise HTTPException(status_code=500, detail="failed_to_create_token")

    data = resp.json()
    token = data.get("data", {}).get("token")
    if not token:
        raise HTTPException(status_code=500, detail="failed_to_create_token")

    return {"token": token}
